src/circuit/gadget/ecc/chip/mul/incomplete.rs

```diff
@@ -0,0 +1,296 @@
+            let lambda1_cur = meta.query_advice(self.lambda1, Rotation::cur());
+            let lambda2_cur = meta.query_advice(self.lambda2, Rotation::cur());
+            let lambda1_next = meta.query_advice(self.lambda1, Rotation::next());
+            let lambda2_next = meta.query_advice(self.lambda2, Rotation::next());
+
+            // The current bit in the scalar decomposition, k_i = z_i - 2⋅z_{i+1}.
+            // Recall that we assigned the cumulative variable `z_i` in descending order,
+            // i from n down to 0. So z_{i+1} corresponds to the `z_prev` query.
+            let k = z_cur - Expression::Constant(pallas::Base::from_u64(2)) * z_prev;
+
+            // y_{A,i} = (λ_{1,i} + λ_{2,i})
+            //           * (x_{A,i} - (λ_{1,i}^2 - x_{A,i} - x_{P,i})) / 2
+            let y_a_cur = (lambda1_cur.clone() + lambda2_cur.clone())
+                * (x_a_cur.clone()
+                    - (lambda1_cur.clone() * lambda1_cur.clone()
+                        - x_a_cur.clone()
+                        - x_p_cur.clone()))
+                * pallas::Base::TWO_INV;
+
+            // y_{A,i+1} = (λ_{1,i+1} + λ_{2,i+1})
+            //           * (x_{A,i+1} - (λ_{1,i+1}^2 - x_{A,i+1} - x_{P,i+1})) / 2
+            let y_a_next = (lambda1_next.clone() + lambda2_next)
+                * (x_a_next.clone()
+                    - (lambda1_next.clone() * lambda1_next - x_a_next.clone() - x_p_next.clone()))
+                * pallas::Base::TWO_INV;
+
+            // Check booleanity of decomposition.
+            let bool_check = k.clone() * (k.clone() + Expression::Constant(-pallas::Base::one()));
+
+            // The base used in double-and-add remains constant. We check that its
+            // x- and y- coordinates are the same throughout.
+            let x_p_check = x_p_cur.clone() - x_p_next;
+            let y_p_check = y_p_cur.clone() - y_p_next;
+
+            // λ_{1,i}⋅(x_{A,i} − x_{P,i}) − y_{A,i} + (2k_i - 1) y_{P,i} = 0
+            let poly1 = lambda1_cur.clone() * (x_a_cur.clone() - x_p_cur.clone()) - y_a_cur.clone()
+                + (k * pallas::Base::from_u64(2) + Expression::Constant(-pallas::Base::one()))
+                    * y_p_cur;
+
+            // (λ_{1,i} + λ_{2,i})⋅(x_{A,i} − (λ_{1,i}^2 − x_{A,i} − x_{P,i})) − 2y_{A,i}) = 0
+            let poly2 = {
+                let lambda_neg = lambda1_cur.clone() + lambda2_cur.clone();
+                let lambda1_expr =
+                    lambda1_cur.clone() * lambda1_cur.clone() - x_a_cur.clone() - x_p_cur.clone();
+                lambda_neg * (x_a_cur.clone() - lambda1_expr)
+                    - Expression::Constant(pallas::Base::from_u64(2)) * y_a_cur.clone()
+            };
+
+            // λ_{2,i}^2 − x_{A,i+1} −(λ_{1,i}^2 − x_{A,i} − x_{P,i}) − x_{A,i} = 0
+            let poly3 = lambda2_cur.clone() * lambda2_cur.clone()
+                - x_a_next.clone()
+                - (lambda1_cur.clone() * lambda1_cur)
+                + x_p_cur;
```

Comment:
```suggestion
            let secant_line =
                lambda2_cur.clone().square() - x_a_next.clone() - lambda1_cur.square() + x_p_cur;
```

Non-blocking: We should also make the `x_a_cur` elimination simplification in the Sinsemilla gate (which I think is what would remove the need for an `x_r` helper). This should happen in a separate PR.

Reply:
I will make a book PR for that.

Reply:
Is this what you meant? https://hackmd.io/iOw7-HpFQY6dPF1aFY8pAw?both#Optimized-Sinsemilla-gate

Note that there is still a reference to x_{A,i} in each step.

Reply:
We established that this doesn't improve performance because x<sub>R,i</sub> still needs to be computed for Y<sub>A,i</sub>. (It's possible to save one addition by defining a temporary as λ<sub>1,i</sub><sup>2</sup> - x<sub>P,i</sub> but that's neither here nor there; we should take the simpler option and avoid changes at this stage.)